Use core connections API in processing database algorithms instead of Python versions

nyalldawson opened the pull request:
Ports the Postgres and Spatialite algorithms to the core database connection API instead of Python versions of these.

The python classes are basically 1:1 duplicates of the core classes, without unit tests and without a common API.

Part of the ongoing Processing database handling cleanup work, but also adds nice API for creation of spatial indexes which can be later used as part of the db manager cleanup/porting efforts.

Refs NRCan Contract#3000707093

nyalldawson committed:
Add API for creating spatial indexes to QgsAbstractDatabaseProviderConnection

nyalldawson committed:
Implement spatial index creation for Postgis connections

nyalldawson committed:
[processing] Use core connections API instead of Python GeoDB class
for postgres connection handling

nyalldawson committed:
Create spatial index support for spatialite connections

nyalldawson committed:
[processing] Use core connections API for Spatialite algorithms instead
of Python reimplementation

nyalldawson committed:
Remove unused spatialite python connection class

nyalldawson committed:
Raise an exception when connection is not specified

nyalldawson committed:
Add CreateSpatialIndex capability for GPKG connection

elpaso commented on python/plugins/processing/algs/gdal/ogr2ogrtopostgislist.py:
```diff
@@ -244,7 +255,7 @@ def getConsoleCommands(self, parameters, context, feedback, executing=True):
         arguments.append('PostgreSQL')
         arguments.append('PG:"')
-        for token in uri.connectionInfo(executing).split(' '):
+        for token in QgsDataSourceUri(uri).connectionInfo(executing).split(' '):
```
Shouldn't use `decodeUri` here?

nyalldawson replied:
Probably, but I'm reluctant to change anything I don't have to here :smiley:

elpaso reviewed (approved):
Maybe you can use `decodeUri` in some places instead of splitting on ` ` but I didn't check if it's really appropriate in this case.

nyalldawson committed:
Add capability for determining whether a spatial index exists (implemented
for GPKG, spatialite) and deleting a spatial index (implemented for GPKG)
to database connection

nyalldawson committed:
Port last postgis alg to connections API

nyalldawson committed:
Drop postgis GeoDB class from Processing

This class is unused and is a near 1:1 copy of the core connections API
+ db manager classes

nyalldawson committed:
Add SpatialIndexExists and DeleteSpatialIndex capabilities for postgres connection

nyalldawson committed:
Also implement hasSpatialIndex for postgres and spatialite

nyalldawson committed:
Fix warnings

nirvn commented:
@nyalldawson , nice work.
